### PR TITLE
KAFKA-3016: phase-2. stream join implementations

### PR DESCRIPTION
**Pull request by ymatsuda**
@guozhangwang

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Specifies that records of the same key are joinable if their timestamp stamps are within
+     * timeDifference.
+     *
+     * @param timeDifference
+     * @return
+     */
+    public JoinWindowSpec within(long timeDifference) {
+        return new JoinWindowSpec(name, timeDifference, timeDifference, retention, segments);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java:51]: Do we need to create a new JoinWindowSpec? Could we make the variables as not final and just override these and return "this"?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Specifies that records of the same key are joinable if their timestamp stamps are within
+     * timeDifference, and if the timestamp of a record from the secondary stream is
+     * is later than or equal to the timestamp of a record from the first stream.
+     *
+     * @param timeDifference
+     * @return
+     */
+    public JoinWindowSpec after(long timeDifference) {
+        return new JoinWindowSpec(name, 0L, timeDifference, retention, segments);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java:75]: Also with `JoinWindowSpec.of(...).before(5L).after(3L)`, then the before value is omitted. Is this the expected behavior?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Specifies the retention period of windows
+     * @param retentionPeriod
+     * @return
+     */
+    public JoinWindowSpec retentionPeriod(long retentionPeriod) {
+        return new JoinWindowSpec(name, before, after, retentionPeriod, segments);
+    }
+
+    public JoinWindowSpec segments(int segments) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java:87]: Personally I feel this API should not be exposed to the users, but only to "developers".

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 4]
```diff
@@ -239,6 +235,135 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
...
+            Deserializer<K> keyDeserialier,
+            Deserializer<V> thisValueDeserialzier,
+            Deserializer<V1> otherValueDeserialzier,
+            boolean outer) {
+
+        Set<String> allSourceNodes = ensureJoinableWith((AbstractStream<K>) other);
+
+        RocksDBWindowStoreSupplier<K, V> thisWindow =
+                new RocksDBWindowStoreSupplier<>(
+                        joinWindowSpec.name + "-1",
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:289]: nit: "-innerRelation" and "-outerRelation" for better debuggability?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamJoinWindow.java [hunk 5]
```diff
@@ -17,52 +17,42 @@
...
-                context().forward(key, value);
-            }
+            window = (WindowStore<K, V>) context.getStateStore(windowName);
         }
 
         @Override
-        public void close() {
-            window.close();
+        public void process(K key, V value) {
+            context().forward(key, value);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamJoinWindow.java:53]: Should we first put into window then forward?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 6]
```diff
@@ -239,6 +235,135 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
...
+                        joinWindowSpec.retention,
+                        joinWindowSpec.segments,
+                        new Serdes<>("", keySerialzier, keyDeserialier, otherValueSerialzier, otherValueDeserialzier),
+                        null);
+
+        KStreamJoinWindow<K, V1> otherWindowedStream = new KStreamJoinWindow<>(otherWindow.name());
+        KStreamKStreamLeftJoin<K, R, V, V1> joinThis = new KStreamKStreamLeftJoin<>(otherWindow.name(), joiner);
+
+        String otherWindowStreamName = topology.newName(WINDOWED_NAME);
+        String name = topology.newName(LEFTJOIN_NAME);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: nit: "name" => "joinThisName"

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.kafka.streams.kstream.ValueJoiner;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.apache.kafka.streams.state.WindowStore;
+
+import java.util.Iterator;
+
+class KStreamKStreamLeftJoin<K, R, V1, V2> implements ProcessorSupplier<K, V1> {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java]: Is the only difference with KStreamKStreamJoin regarding "outer" parameter? In this case can we call 

`KStreamKStreamJoin<K, R, V, V1> joinThis = new KStreamKStreamJoin<>(otherWindow.name(), joiner, true);`

in line 355 of KStreamImpl?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java (hunk 2 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java:75]: That was my intention. Do you think allowing both before and after useful? If so, I can change it.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 4 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:289]: How about "this" and "other" to be consistent with other places?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java (hunk 2 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindowSpec.java:75]: Maybe, I am not sure. Maybe just leave the door open for now?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 4 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:289]: Sounds good.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 6 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamJoinWindow.java (hunk 5 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamJoinWindow.java:53]: It shouldn't make any practical difference, but I think it is safer to update the state after all downstream processing is performed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java (hunk 7 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java]: You are right. Good catch!

**Comment by guozhangwang** [comment]
LGTM. Merging to trunk.

**Pull request closed without merging**